server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -1475,7 +1485,37 @@ private Set<DataSegment> createNewIdsForAppendSegments(
           entry.getValue(),
           appendVersionToSegments
       );
-      for (Map.Entry<Interval, Set<DataSegment>> upgradeEntry : segmentsToUpgrade.entrySet()) {
+      Map<Interval, Set<DataSegment>> filteredSegmentsToUpgrade = new HashMap<>();
+      for (Interval interval : segmentsToUpgrade.keySet()) {
+        filteredSegmentsToUpgrade.put(interval, new HashSet<>());
+        for (DataSegment segment : segmentsToUpgrade.get(interval)) {
+          boolean include = true;
+          if (segmentIdToUpgradeVersions.containsKey(segment.getId().toString())) {
+            for (SegmentIdWithShardSpec upgradedSegment : segmentIdToUpgradeVersions.get(segment.getId().toString())) {
+              if (upgradedSegment.getVersion().equals(upgradeVersion)) {
+                upgradedSegments.add(
+                    new DataSegment(
+                        upgradedSegment.asSegmentId(),
+                        segment.getLoadSpec(),
+                        segment.getDimensions(),
+                        segment.getMetrics(),
+                        upgradedSegment.getShardSpec(),
+                        null,
+                        segment.getBinaryVersion(),
+                        segment.getSize()
+                    )
+                );
+                include = false;
+                break;
+              }
+            }
+          }
+          if (include) {
+            filteredSegmentsToUpgrade.get(interval).add(segment);
+          }
+        }
+      }
```

Comment:
can you please some comments to describe what are you trying to do here? I don't follow.